added check for fee recipient per validator and added unit tests

AbolareRoheemah opened the pull request:
## Issue Addressed
Addresses #5403 

## Proposed Changes
 - Added `check_fee_recipient()` method to validate individual validators
 - Added `check_all_fee_recipients()` to validate all validators on startup
- Validator client now fails to start if any enabled validator lacks a fee recipient and no global flag is used.
 - Added Clear error messages to guide users on how to fix the issue
 - Added unit tests

## Additional Info
The images show attempts to run the VC: 1. without fee-recipient set for a validator and no global flag 2. Only global flag used. No individual fee-recipient set
<img width="800" height="347" alt="Screenshot 2025-11-23 at 20 09 08" src="https://github.com/user-attachments/assets/c3e24573-50e5-4617-8df5-d330a71e93f2" />
<img width="792" height="273" alt="Screenshot 2025-11-23 at 20 10 21" src="https://github.com/user-attachments/assets/fac4c3bf-dd2f-4274-aae2-d6eea8684cd6" />

AbolareRoheemah committed:
added check for fee recipient per validator and added unit tests

eserilev commented on common/account_utils/src/validator_definitions.rs:
```diff
@@ -211,6 +211,26 @@ impl ValidatorDefinition {
+
+        // If nothing set
+        Err(format!(
+            "Validator {} is missing `suggested_fee_recipient`!\n\n\
```
Can we update the error message to

```
Validator {} is missing a `suggested_fee_recipient`. Fix this by adding a `suggested_fee_recipient` in your `validator_definitions.yml` or by supplying a fallback fee recipient via the `--suggested-fee-recipient` flag.
```

We usually avoid `\n`. We can use `\` to continue the string literal across multiple lines

eserilev commented on common/account_utils/src/validator_definitions.rs:
```diff
@@ -211,6 +211,26 @@ impl ValidatorDefinition {
         })
     }
+
+    pub fn check_fee_recipient(&self, global_fee_recipient: Option<Address>) -> Result<(), String> {
```
I think for better UX, we should iterate across all fee recipients, build a list of ones missing fee recipients and then log an error to the user. This way they can see the full list of missing fee recipients all at once.

eserilev commented on common/account_utils/src/validator_definitions.rs:
```diff
@@ -409,6 +429,34 @@ impl ValidatorDefinitions {
+            if missing > 0 {
+                tracing::warn!(
+                    "Global --suggested-fee-recipient is being used for {} validator(s). \
+                        Consider setting it in validator_definitions.yml for each one.",
```
```suggestion
                    "The fallback --suggested-fee-recipient is being used for {} validator(s). \
                        Consider setting the fee recipient for each validator individually via `validator_definitions.yml`.",
```

eserilev commented on common/account_utils/src/validator_definitions.rs:
```diff
@@ -409,6 +429,34 @@ impl ValidatorDefinitions {
+                .filter(|d| d.enabled && d.suggested_fee_recipient.is_none())
+                .count();
+            if missing > 0 {
+                tracing::warn!(
```
```suggestion
                tracing::info!(
```

eserilev reviewed (changes requested):
Thanks for this! The added tests are very much appreciated. I think we can make some slight tweaks here to improve the UX a bit.

AbolareRoheemah committed:
error now lists all missing validators, using info instead of warn and modified code comment

AbolareRoheemah commented:
> Thanks for this! The added tests are very much appreciated. I think we can make some slight tweaks here to improve the UX a bit.

Hello @eserilev. Much thanks for the review and the kind words. 🙏

I thought to explain the changes I made with this last push. The `check_fee_reciepient` function now returns the public key of the validator with missing fee recipient. This is done to make collecting all missing validators easy.
Then `check_all_fee_recipients` now gathers all validators missing a fee recipient and reports them together in a single error message (comma-separated list).

I also modified existing tests to work with the new `check_fee_reciepient` return type and added a test (check_all_fee_recipients_reports_all_missing) that verifies multiple missing validators are all reported in a single error.

AbolareRoheemah committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into make-fee-receipient-mandatory

eserilev reviewed (approved):
LGTM!

eserilev commented:
Hi @AbolareRoheemah can you please run `cargo fmt --all` to fix the CI issues when you get the chance

chong-he commented on common/account_utils/src/validator_definitions.rs:
```diff
@@ -682,4 +739,250 @@ mod tests {
+        };
+
+        // Should return None since global fee recipient is set
+        let global_fee = Some(Address::from_str("0xa2e334e71511686bcfe38bb3ee1ad8f6babcc03d").unwrap());
```
This is just a nit for better clarity 
```suggestion
        let global_fee_recipient = Some(Address::from_str("0xa2e334e71511686bcfe38bb3ee1ad8f6babcc03d").unwrap());
```

AbolareRoheemah replied:
Thanks. It does make things clearer. 👍

chong-he commented on common/account_utils/src/validator_definitions.rs:
```diff
@@ -682,4 +739,250 @@ mod tests {
+        // Should return None since global fee recipient is set
+        let global_fee = Some(Address::from_str("0xa2e334e71511686bcfe38bb3ee1ad8f6babcc03d").unwrap());
+        let check_result = def.check_fee_recipient(global_fee);
+        assert!(check_result.is_none());
```
For succinctness, we can combine this with the test `fee_recipient_check_fails_when_missing()` to avoid repeatedly defining the same fields. We can revise the test name a bit, then add the checks in the same tests

Edit: I believe we can even go "all out" for succinctness, to define a general `ValidatorDefinitions`, then for each test, update the necessary fields for each test purpose, and add an assert right below, i.e., instead of having multiple tests, we can have just 1 or 2 test functions, move the test name (and revise a bit) and put it as a comment right above and before the assert. This will be cleaner as well

AbolareRoheemah replied:
Thanks for spotting this. I feel those two tests should be combined. 🤝

About having just 1 or 2 test functions, I actually tried compressing the tests into 3 functions, and while it makes things more compact, the functions appeared not so comprehensible due to how long they became. So I think I prefer to separate test concerns as much as possible.

chong-he replied:
Let's combine those 2 tests at least

To me, having a longer function is ok as long as the tests are clearly defined / commented above it. It is nicer to have fewer lines that can cover the tests (e.g., avoid defining the fields repeatedly) looks cleaner.

AbolareRoheemah replied:
> For succinctness, we can combine this with the test `fee_recipient_check_fails_when_missing()` to avoid repeatedly defining the same fields. We can revise the test name a bit, then add the checks in the same tests

Yes. The two tests you highlighted have been combined in the update I pushed yesterday. 🫡

chong-he commented on common/account_utils/src/validator_definitions.rs:
```diff
@@ -410,6 +420,52 @@ impl ValidatorDefinitions {
+            
+            return Err(format!(
+                "The following validators are missing a `suggested_fee_recipient`: {}. \
+                 Fix this by adding a `suggested_fee_recipient` in your \
```
I feel that `the validator_definitions.yml` reads better here, but feel free to take it or leave it
```suggestion
                 Fix this by adding a `suggested_fee_recipient` in the \
```

chong-he commented on common/account_utils/src/validator_definitions.rs:
```diff
@@ -410,6 +420,52 @@ impl ValidatorDefinitions {
+            if count > 0 {
+                info!(
+                    "The fallback --suggested-fee-recipient is being used for {} validator(s). \
+                     Consider setting the fee recipient for each validator individually via `validator_definitions.yml`.",
```
This is about "reducing the severity" of the log, because if users only use the `suggested-fee-recipient` flag, this log will appear every time a VC starts, so it could be a bit of a warn thingy.

This is a my two cent, keep to hear from the author and other developers about this. (e.g., maybe even shift to debug log)?

```suggestion
                     You may alternatively set the fee recipient for each validator individually via `validator_definitions.yml`.",
```

AbolareRoheemah replied:
For this, I previously used a warn, and @eserilev suggested I change it to an info log.  I don't mind changing it to a debug or something.

chong-he replied:
Let's wait to hear from Eitan and see what he says. 

I think even removing this log looks good to me (or move to debug), because the `suggested_fee_recipient` is something sensitive to users and having a new INFO log there could be a surprise and users may start to questions or panic if they should define in the yml file

eserilev replied:
yeah debug or removing the log seems fine to me.

AbolareRoheemah replied:
Noted

AbolareRoheemah replied:
Done @eserilev @chong-he

AbolareRoheemah replied:
Done @chong-he @eserilev

chong-he reviewed (commented):
Looks pretty good to me. I have also tested manually and confirm that it's working. 

Some comments below.

AbolareRoheemah committed:
ran "cargo fmt --all"

AbolareRoheemah committed:
merged fee_recipient_check_fails_when_missing test with fee_recipient_check_passes_with_global_flag into 1 test "fee_recipient_check_enabled_validator_cases"

AbolareRoheemah commented:
> Looks pretty good to me. I have also tested manually and confirm that it's working.
> 
> Some comments below.

Thank for your review @chong-he and sorry for the delayed response.

AbolareRoheemah committed:
updated error text

AbolareRoheemah commented:
> Hi @AbolareRoheemah can you please run `cargo fmt --all` to fix the CI issues when you get the chance

Done! Sorry for the late response. 🙇‍♀️

AbolareRoheemah committed:
replaced info log with lower level debug

chong-he commented:
> > Hi @AbolareRoheemah can you please run `cargo fmt --all` to fix the CI issues when you get the chance
> 
> Done! Sorry for the late response. 🙇‍♀️



> > Hi @AbolareRoheemah can you please run `cargo fmt --all` to fix the CI issues when you get the chance
> 
> Done! Sorry for the late response. 🙇‍♀️

Hi, you also need to change the `info!` to `debug!` if you want the log to display as `DEBUG`. Alternatively, for simplicity we can remove the log altogether and also reduce quite a few lines in the code

AbolareRoheemah commented:
> Hi, you also need to change the `info!` to `debug!` if you want the log to display as `DEBUG`.

I can't believe I changed the import from to debug and didn't change the actual implementation. 🤦‍♀️

> Alternatively, for simplicity we can remove the log altogether and also reduce quite a few lines in the code

For this, are you suggesting there's no need to inform users when the --suggested-fee-recipient fallback is being used? I would appreciate if you could clarify further. 🙏 @chong-he

chong-he commented:
> For this, are you suggesting there's no need to inform users when the --suggested-fee-recipient fallback is being used? I would appreciate if you could clarify further. 🙏 @chong-he

Yes, we can simplify to not logging anything, that can make it cleaner
but if you want to keep it I am good with it as well, we log as debug. I am fine either way

AbolareRoheemah committed:
replaced info with debug

AbolareRoheemah commented:
> Yes, we can simplify to not logging anything, that can make it cleaner but if you want to keep it I am good with it as well, we log as debug. I am fine either way

Thanks for clarifying @chong-he. 

So I kept the debug log because I think it could possibly prevent accidental misconfiguration or missed rewards or just encourage or inform users of "best practices."

chong-he reviewed (approved):
Looks good now, thanks for the PR

jimmygchen commented:
@chong-he 

One thing worth considering: validators added at runtime via the HTTP API (POST `/lighthouse/validators/*` and the `web3signer/remotekeys` endpoints) don't go through this check. The `suggested_fee_recipient` field is optional in all those request bodies, so a validator can still be added without one and silently lose rewards to 0x00..01.

Not necessarily a blocker for this PR, but worth deciding whether runtime-added validators should get the same protection (either by validating in the API handlers, or noting it as a follow-up).

chong-he commented:
> @chong-he
> 
> One thing worth considering: validators added at runtime via the HTTP API (POST `/lighthouse/validators/*` and the `web3signer/remotekeys` endpoints) don't go through this check. The `suggested_fee_recipient` field is optional in all those request bodies, so a validator can still be added without one and silently lose rewards to 0x00..01.
> 
> Not necessarily a blocker for this PR, but worth deciding whether runtime-added validators should get the same protection (either by validating in the API handlers, or noting it as a follow-up).

Good point. 

If we validate in the API handlers, should it prevent importing the key if no `suggested_fee_recipient` is provided? My thinking is no because the KeyManager API or `/lighthouse/validators/keystore` has no requirement for this. Maybe add a warning during importing via HTTP API if it is not provided? Then, it will be checked upon a restart. What do you think?
